src/ranged.cpp

```diff
@@ -1323,714 +1463,599 @@ std::vector<aim_type> Character::get_aim_types( const item &gun ) const
+            if( !attack_was_confirmed && !confirm_non_enemy_target() ) {
                 continue;
             }
-            break;
-        } else if( action == "CENTER" ) {
-            pc.view_offset = tripoint_zero;
-            dst = src;
-            set_last_target( dst );
-            ret.clear();
-        } else if( action == "TOGGLE_SNAP_TO_TARGET" ) {
-            if( snap_to_target ) {
-                snap_to_target = false;
-                pc.view_offset = dst - pc.pos();
+
+            if( action_aim_and_shoot( pc, action ) ) {
+                loop_exit_code = ExitCode::Fire;
             } else {
-                snap_to_target = true;
+                timed_out_action = action;
+                loop_exit_code = ExitCode::Timeout;
             }
-        } else if( action == "QUIT" ) { // return empty vector (cancel)
-            ret.clear();
-            pc.last_target_pos = cata::nullopt;
             break;
-        } else if( action == "zoom_in" ) {
-            g->zoom_in();
-        } else if( action == "zoom_out" ) {
-            g->zoom_out();
-        } else if( action == "MOUSE_MOVE" || action == "TIMEOUT" ) {
-            tripoint edge_scroll = g->mouse_edge_scrolling_terrain( ctxt );
-            if( edge_scroll == tripoint_zero ) {
-                redraw = false;
-            } else {
-                if( action == "MOUSE_MOVE" ) {
-                    edge_scroll *= 2;
-                }
-                pc.view_offset += edge_scroll;
-                if( snap_to_target ) {
-                    dst += edge_scroll;
-                }
-            }
         }
+    } // for(;;)
 
-        int new_dx = dst.x - src.x;
-        int new_dy = dst.y - src.y;
+    set_view_offset( pc, saved_view_offset );
 
-        // Make player's sprite flip to face the current target
-        if( !tile_iso ) {
-            if( new_dx > 0 ) {
-                g->u.facing = FD_RIGHT;
-            } else if( new_dx < 0 ) {
-                g->u.facing = FD_LEFT;
-            }
-        } else {
-            if( new_dx >= 0 && new_dy >= 0 ) {
-                g->u.facing = FD_RIGHT;
-            }
-            if( new_dy <= 0 && new_dx <= 0 ) {
-                g->u.facing = FD_LEFT;
-            }
+    switch( loop_exit_code ) {
+        case ExitCode::Abort: {
+            traj.clear();
+            break;
         }
-    } while( true );
-
-    pc.view_offset = old_offset;
-
-    if( ret.empty() ) {
-        return ret;
-    }
-
-    set_last_target( ret.back() );
-
-    const auto lt_ptr = pc.last_target.lock();
-    if( npc *const guy = dynamic_cast<npc *>( lt_ptr.get() ) ) {
-        if( !guy->guaranteed_hostile() ) {
-            // TODO: get rid of this. Or combine it with effect_hit_by_player
-            guy->hit_by_player = true; // used for morale penalty
+        case ExitCode::Fire: {
+            on_target_accepted( pc, true );
```

Comment:
This true should be some other condition to prevent that.

Reply:
Yeah, this true was supposed to be that removed condition :/